python_modules/libraries/dagster-components/dagster_components/lib/sling_replication_collection/component.py

```diff
@@ -18,37 +18,72 @@
 from dagster_components.core.schema.objects import (
     AssetAttributesModel,
     AssetSpecTransformModel,
-    OpSpecBaseModel,
+    OpSpecModel,
+    ResolvableModel,
 )
+from dagster_components.core.schema.resolver import TemplatedValueResolver
 from dagster_components.utils import ResolvingInfo, get_wrapped_translator_class
 
 
-class SlingReplicationParams(BaseModel):
+@record
+class SlingReplicationSpec:
+    relative_path: str
+    op_spec: OpSpecModel
+    translator: DagsterSlingTranslator
+
+
+class SlingReplicationParams(ResolvableModel[SlingReplicationSpec]):
     path: str
-    op: Optional[OpSpecBaseModel] = None
+    op: Optional[OpSpecModel] = None
     asset_attributes: Annotated[
         Optional[AssetAttributesModel], ResolvableFieldInfo(required_scope={"stream_definition"})
     ] = None
 
+    def resolve(self, resolver: TemplatedValueResolver) -> SlingReplicationSpec:
+        return SlingReplicationSpec(
+            relative_path=resolver.resolve_obj(self.path),
+            op_spec=self.op.resolve(resolver) if self.op else OpSpecModel(),
+            translator=get_wrapped_translator_class(DagsterSlingTranslator)(
+                resolving_info=ResolvingInfo(
+                    "stream_definition", self.asset_attributes or AssetAttributesModel(), resolver
+                ),
+            ),
+        )
+
 
-class SlingReplicationCollectionParams(BaseModel):
+class SlingReplicationCollectionParams(ResolvableModel["SlingReplicationCollection"]):
     sling: Optional[SlingResource] = None
     replications: Sequence[SlingReplicationParams]
     transforms: Optional[Sequence[AssetSpecTransformModel]] = None
 
+    def resolve(
+        self, resolver: TemplatedValueResolver
+    ) -> tuple[
+        SlingResource,
+        Sequence[SlingReplicationSpec],
+        Sequence[Callable[[Definitions], Definitions]],
+    ]:
+        return (
+            self.sling if self.sling else SlingResource(),
+            [replication.resolve(resolver) for replication in self.replications],
+            [transform.resolve(resolver) for transform in self.transforms]
+            if self.transforms
+            else [],
+        )
+
```

Comment:
Ok I am still confused here. Shouldn't the return value of this `resolve` be the same as the generic argument to the base class (`"SlingReplicationCollection"`).

Reply:
I think some sort of dedicated type for that tuple would be clarifying.

Reply:
Ah this was a mistake -- I mentioned in slack that returning an instance of a component class from the `resolve()` method was problematic. When I was refactoring this to avoid doing that, I forgot to update the type signature on the class

Just updated!